### ISSUE TITLE
Error processing PDB file, java.lang.NullPointerException

### ISSUE THREAD

**Issue by JosephClampett** [NationalSecurityAgency/ghidra#3484]
**Describe the bug**
A clear and concise description of the bug.

**To Reproduce**
1. Import PE file.
2. Choose not to start auto-analysis.
3. Load PDB using Universal.
4. See error as in log file. java.lang.NullPointerException.

**Expected behavior**
The PDB should be correctly loaded so that Auto Analysis can proceed.

**Attachments**
[application.log](https://github.com/NationalSecurityAgency/ghidra/files/7316621/application.log)
Unfortunately I can't attach the exe and PDB because they are copyrighted but I can send them directly.

**Environment:**
 - OS: Windows 11 21H2 (22000.194)
 - Java Version: openjdk version "11" 2018-09-25
 - Ghidra Version: 10.0.4
 - Ghidra Origin: https://github.com/NationalSecurityAgency/ghidra/releases

**Additional context**
This PDB is 1 GB in size.

**Comment by JosephClampett** [comment]
Got it. Do you want me to put in a check and build from source in the meantime and/or send you the exe and PDB via Drive?

**Comment by JosephClampett** [comment]
I have not, no. I did get it working with the old PDB tool (though I had to fix a few bugs first), though, by forcing it to create an enormous XML first and then loading that in Ghidra.

**Comment by ghizard** [comment]
@JosephClampett

Thanks for providing the diagnostic information back.

After studying the code a bit, I'd like us to do two things.  One is to put code into place to prevent the NPE.  After that is tested, then I'd like to take a leap to a possible solution, but I'm really leaping, so it could be wrong, but it might be the ultimate issue.

First, replace the `ArrayTypeApplier` code I gave you before for just below line 121 with the following:
```
		if (underlyingDataType == null) {
			Long v = longUnderlyingSize;
			underlyingDataType = Undefined.getUndefinedDataType(v.intValue());
			String msg = "PDB Type index " + index + ":\n   Null underlying data type for " +
				underlyingTypeApplier.getClass().getSimpleName() + ":\n      " +
				underlyingTypeApplier + "\n   Using " + underlyingDataType;
			Msg.warn(this, msg);
		}
```
Then see if you can test it with your program to see if Ghidra no longer NPEs and at the same time check that this new code does not cause problems.  (We could uncover "other" problems in your program, however, once we can process through this one.)  I'd like you to confirm that the warning looks OK too.

Once that is in place, add the following code to EnumTypeApplier, between the `getName()` method on line 108 and the `getUnderlyingTypeApplier()` method on line 112:
```
	@Override
	DataType getDataType() {
		getOrCreateEnum();
		return dataType;
	}
```

I hope this fixes the issue, but if not, we can take additional steps to hone in on the solution.  But if it works, then the new message that we added in ArrayTypeApplier should no longer get logged.

**Comment by JosephClampett** [comment]
Only adding the first change causes a OutOfMemoryError.
[application.log](https://github.com/NationalSecurityAgency/ghidra/files/7464927/application.log)

**Comment by ghizard** [comment]
@JosephClampett

I see that your application log had:
```
2021-11-02 22:06:30 WARN  (ArrayTypeApplier) PDB Type index 884356:
   Null underlying data type for EnumTypeApplier:
      enum Black::Entity::Map::ArchetypeFixId<9,unsigned,hasuniquename> {public: PARAM_NAV_ARCHETYPE_BUDDY=16839467,public: PARAM_NAV_ARCHETYPE_ENEMY_1A_JUMP=16956324,public: PARAM_NAV_ARCHETYPE_BUDDY_CHOCOBO=16936330,public: PARAM_NAV_ARCHETYPE_ENEMY_1B_JUMP=16956365,public: PARAM_NAV_ARCHETYPE_ENEMY_2B_JUMP=17004384,public: PARAM_NAV_ARCHETYPE_ENEMY_3A_JUMP=16956367,public: PARAM_NAV_ARCHETYPE_ENEMY_2A_JUMP=16956366,public: PARAM_NAV_ARCHETYPE_ENEMY_3B_JUMP=16956368,public: PARAM_NAV_ARCHETYPE_ENEMY_4A_JUMP=16956370} 
   Using undefined4
```
and later had the
```
2021-11-02 22:08:17 ERROR (SwingExceptionHandler) Error: Uncaught Exception! 
Memory: free=2408998928 max=8575254528 total=8575254528 java.lang.OutOfMemoryError: Java heap space
```

This makes be believe that the first bit of code succeeded, but leaves us with two things to do:  Try the second fix to see if the "Null underlying" error goes away, then try to work on the memory situation.

If you can try the second fix (EnumTypeApplier) before working on the memory issue, it can at least help me determine if I need to do more investigation on this issue (and possibly not have to wait hours or (hopefully not) days to see if the second fix works).

To try to address the memory...
First, Ghidra can have issues with big programs and we keep pushing the limits of its ability.  This initial "PDB Universal" design is far from working well on large programs/PDBs, and we've had recent improvements in PDB computation time on this, but still have lots of room to improve memory performance.  To get this computational performance improvement, you need to be working off of master (10.1), which I believe you are doing.  For my large 340MB PDB, computation time has dropped from 4.5 hours to 19 minutes.  But I've run from Eclipse and had set my JVM to have much more memory available to it.  I've set it arbitrarily high at 40GB, though it never used that much.  I don't know how much would be needed for your 1GB PDB, and we might not be successful at all on your 32GB system until other PDB memory improvements are worked out.

If you are not running out of Eclipse, but from a launch script, you should be able to set the JVM heap size in GhidraRun.bat (Windows) or GhidraRun (Linux) by uncommenting and setting a high MAXMEM value (but probably not set to system max).  If you are running out of Eclipse, you would have to set a -Xmx value in the Default VM ARguments of the Installed JREs configurations.

**Comment by ghizard** [comment]
To be clear about your last comment, are you saying the "Null underlying" message still appeared in the log after the EnumTypeApplier change was made?

**Comment by JosephClampett** [comment]
Sorry. I was saying that the memory issue is unsurprisingly still there. The second change fixes the "Null underlying." Going back to your initial question - I was actually using 10.0.4 so I will upgrade first.

**Comment by ghizard** [comment]
That's awesome about the "Null underlying" going away with the change in EnumTypeApplier.  I will try to get these first two changes put in as the "fix" for this ticket.  We can keep discussing the performance issue (OOME) here, if you'd like, but if there is something other than performance, please enter it as a new ticket so that it can be tracked separately from the NPE that occurred here in the ArrayTypeApplier.

**Comment by JosephClampett** [comment]
Yep let's address the memory issues but I'll check it out in about a week.
